tests/pipeline/controlflowanalysis/test_readability_based_refinement.py

```diff
@@ -2077,3 +2079,121 @@ def test_declaration_listop(self, ast_call_for_loop):
+                if(b < 0) {
+                    b = b + 2
+                    continue
+                }
+                b = b + 1
+            }
+            a = a + 1
+        }
+        """
+        true_value = LogicCondition.initialize_true(context := LogicCondition.generate_new_context())
+        ast = AbstractSyntaxTree(
+            root := SeqNode(true_value),
+            condition_map={
+                logic_cond("x1", context): Condition(OperationType.less, [Variable("a"), Constant(5)]),
+                logic_cond("x2", context): Condition(OperationType.less, [Variable("b"), Constant(10)]),
+                logic_cond("x3", context): Condition(OperationType.less, [Variable("b"), Constant(0)])
+            }
+        )
+
+        true_branch = ast._add_code_node(
+            [
+                Assignment(Variable("b"), BinaryOperation(OperationType.plus, [Variable("b"), Constant(2)])),
+                Continue()
+            ]
+        )
+        if_condition = ast._add_condition_node_with(logic_cond("x2", context), true_branch)
+
+        while_loop_outer = ast.factory.create_while_loop_node(logic_cond("x1", context))
+        while_loop_body_outer = ast.factory.create_seq_node()
+        while_loop_iteration_outer_1 = ast._add_code_node([Assignment(Variable("a"), BinaryOperation(OperationType.plus, [Variable("a"), Variable("b")]))])
+        while_loop_iteration_outer_2 = ast._add_code_node([Assignment(Variable("a"), BinaryOperation(OperationType.plus, [Variable("a"), Constant(1)]))])
+        ast._add_node(while_loop_outer)
+        ast._add_node(while_loop_body_outer)
+
+        while_loop_inner = ast.factory.create_while_loop_node(logic_cond("x1", context))
+        while_loop_body_inner = ast.factory.create_seq_node()
+        while_loop_iteration_inner = ast._add_code_node([Assignment(Variable("b"), BinaryOperation(OperationType.plus, [Variable("b"), Constant(1)]))])
+        ast._add_node(while_loop_inner)
+        ast._add_node(while_loop_body_inner)
+
+        ast._add_edges_from(
+            [
+                (root, while_loop_outer),
+                (while_loop_outer, while_loop_body_outer),
+                (while_loop_body_outer, while_loop_inner),
+                (while_loop_body_outer, while_loop_iteration_outer_1),
+                (while_loop_body_outer, while_loop_iteration_outer_2),
+                (while_loop_inner, while_loop_body_inner),
+                (while_loop_body_inner, if_condition),
+                (while_loop_body_inner, while_loop_iteration_inner)
+            ]
+        )
+
+        WhileLoopReplacer(ast, _generate_options()).run()
+        for loop_node in list(ast.get_loop_nodes_post_order()):
+            if any(node.does_end_with_continue for node in loop_node.body.get_descendant_code_nodes_interrupting_ancestor_loop()):
+                assert isinstance(loop_node, WhileLoopNode)
+            else:
+                assert isinstance(loop_node, ForLoopNode)
```

Comment:
Same as above, the test is not sufficient. You have to check that the outer-loop is replaced in the inner-loop is not replaced. Since you iterate the nodes in post-order, you know which loop is first.
Additionally, you could change the loop-break condition of one and check that the for-loop belongs to the correct update-condition and the other not. But, the order would be sufficient.

Reply:
you should change the condition of the inner loop to `b <= 5`, otherwise it does not match a for-loop